src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,397 @@
+        authorization_base64_header =
+                rd_malloc(authorization_base64_header_size);
+
+        rd_snprintf(authorization_base64_header,
+                    authorization_base64_header_size,
+                    authorization_header,
+                    client_authorization_out.ptr);
+
+        rd_free(client_authorization_in.ptr);
+        rd_free(client_authorization_out.ptr);
+        return authorization_base64_header;
+}
+
+
+/**
+ * @brief Build headers for HTTP(S) requests.
+ *
+ * @returns The result will be responsed in \p *headers.
+ *          If succeed, return rd_true with NULL in \p *error,
+ *          If failed, return rd_false with error message in \p *error.
+ *
+ * @locality Any thread.
+ */
+static rd_bool_t build_headers (rd_kafka_conf_t *conf,
+                                struct curl_slist **headers,
+                                char **error) {
+        static const char *miss_client_id_error_message = "client_id is "
+                "a required field";
+        static const char *miss_client_secret_error_message = "client_secret "
+                "is a required field";
+        static const char *accept_header = "Accept:application/json";
+        const char *client_id;
+        const char *client_secret;
+        char *authorization_base64_header;
+
+        client_id = conf->sasl.oauthbearer.client_id;
+
+        if (client_id == NULL) {
+                *error = rd_malloc(strlen(miss_client_id_error_message)
+                                   + 1);
+                strcpy(*error, miss_client_id_error_message);
+                return rd_false;
+        }
+
+        client_secret = conf->sasl.oauthbearer.client_secret;
+        if (client_secret == NULL) {
+                *error = rd_malloc(strlen(miss_client_secret_error_message)
+                                   + 1);
+                strcpy(*error, miss_client_secret_error_message);
+                return rd_false;
+        }
+
+        authorization_base64_header =
+                build_authorization_base64_header(client_id,
+                                                  client_secret,
+                                                  &*error);
+        if (authorization_base64_header == NULL)
+                return rd_false;
+
+        *headers = curl_slist_append(*headers, accept_header);
+        *headers = curl_slist_append(*headers, authorization_base64_header);
+        rd_free(authorization_base64_header);
+        return rd_true;
+}
+
+
+/**
+ * @brief Implementation of Oauth/OIDC token refresh call back function,
+ *        will receive the json response after HTTP call from token provider,
+ *        then extract the jwt from the json response, and forward it to
+ *        the broker.
+ */
+void rd_kafka_conf_set_oauthbearer_oidc_token_refresh_cb (rd_kafka_t *rk,
+        const char *oauthbearer_config, void *opaque) {
+        static const char *data = "grant_type=client_credentials,scope=%s";
```

Comment:
Predefining these makes the code harder to read below, like the variable 'data' doesnt tell me much.
I recommend using the strings where they are needed instead, and employ the rule of 3 (duplicate something twice is okay, abstract and unify when used 3 times).